common/src/main/java/org/conscrypt/AbstractSessionContext.java

```diff
@@ -199,208 +180,93 @@ public final void setSessionCacheSize(int size)
-        } else {
-            return session;
+        // Let the subclass know.
+        addingSession(session);
+
+        ByteArray key = new ByteArray(id);
+        synchronized (sessions) {
+            sessions.put(key, session);
         }
     }
 
-    @Override
-    public SSLSession getSession(byte[] sessionId) {
+    /**
+     * Called for server sessions only. Retrieves the session by its ID. Overridden by
+     * {@link ServerSessionContext} to
+     */
+    final SslSessionWrapper getSessionFromCache(byte[] sessionId) {
         if (sessionId == null) {
-            throw new NullPointerException("sessionId == null");
+            return null;
         }
-        ByteArray key = new ByteArray(sessionId);
-        SSLSession session;
+
+        // First, look in the in-memory cache.
+        SslSessionWrapper session;
         synchronized (sessions) {
-            session = sessions.get(key);
+            session = sessions.get(new ByteArray(sessionId));
         }
         if (session != null && session.isValid()) {
-            return wrapSSLSessionIfNeeded(session);
+            return session;
         }
-        return null;
-    }
 
-    void putSession(SSLSession session) {
-        byte[] id = session.getId();
-        if (id.length == 0) {
-            return;
-        }
-        ByteArray key = new ByteArray(id);
-        synchronized (sessions) {
-            sessions.put(key, session);
-        }
+        // Not found in-memory - look it up in the persistent cache.
+        return getSessionFromPersistentCache(sessionId);
     }
 
-    private static void log(Throwable t) {
-        System.out.println("Error inflating SSL session: "
-                + (t.getMessage() != null ? t.getMessage() : t.getClass().getName()));
-    }
+    /**
+     * Called when the given session is about to be added. Used by {@link ClientSessionContext} to
+     * update its host-and-port based cache.
+     *
+     * <p>Visible for extension only, not intended to be called directly.
+     */
+    abstract void addingSession(SslSessionWrapper session);
 
-    @Override
-    protected void finalize() throws Throwable {
-        try {
-            NativeCrypto.SSL_CTX_free(sslCtxNativePointer);
-        } finally {
-            super.finalize();
+    /**
+     * Called when a session is about to be removed. Used by {@link ClientSessionContext}
+     * to update its host-and-port based cache.
+     *
+     * <p>Visible for extension only, not intended to be called directly.
+     */
+    abstract void removingSession(SslSessionWrapper session);
+
+    /**
+     * Called for server sessions only. Retrieves the session by ID from the persistent cache.
+     *
+     * <p>Visible for extension only, not intended to be called directly.
+     */
+    abstract SslSessionWrapper getSessionFromPersistentCache(byte[] sessionId);
+
+    /**
+     * Makes sure cache size is < maximumSize.
```

Comment:
The logic here should be identical to the old code. Since they are stored in a `LinkedHashMap`, iteration should be based on insertion order. Or am I missing something?